### DISCUSSION TITLE
Minimize disk writes when using EC

### DISCUSSION THREAD

**Discussion by struthio** [moosefs/moosefs#601] [Q&A]
hi,
Thanks for releasing version 4 for public.
I am trying to change one of my storage classes to use EC, but I have one question.
Currently (with normal SC) I have following SC defined
(S - SSD drives, H - HDD drives)
KEEP: S,2H
ARCHIVE: 2H
So when I upload file it have 3 copies, one fast and two slow. When 'Archive Time Is reached' only one copy (from SSD) is deleted. When somebody wants again to work on file, I usually execute ``mfsclsarchive`` and one additional copy of file to SSD is made.

In new version EC can only be used in Archive, and not in Keep. So for example for setup like:
KEEP: S,2H
ARCHIVE: @1,H
Using ``mfsclsarchive`` on file will cause whole file to be rewritten from EC to 3 regular copies, and after file is archived it will be rewritten from 3 copies to EC format. So each change KEEP/ARCHIVE can cause a lot of disk IO ?

**Comment by chogata** [comment]
Yes, in the scenario you described there will be some extra i/o that would not have happened in your previous configuration. First, one chunk server will prepare parts of a chunk, then these parts will be copied to other chunk servers. The parts are smaller than the whole copy, of course, so less data would be written to each of those other chunk servers, but the total number of i/o operations will be higher in general.

If you want to modify a file, a chunk (not the whole file, just this one chunk) needs to be converted back to copy format, because it's not possible to modify a chunk in EC format. You don't need to use the `mfsclsarchive` first, this will happen automatically. If you just want to read a file, then no conversion is needed.

The EC format is not recommended for files that are modified often - it's to save space on files that are archived, so mostly only read or left alone. We have a special 'per chunk' mode, for big files, that are mostly unmodified, but may have parts that are modified more often (big VM disks, for example). So it all depends on what kind of files you keep on your MooseFS instance and how you work with them.
